### PR TITLE
Support Jsoniter codec

### PR DESCRIPTION
**Pull request by lavrov**

### CODE REVIEW

File: journal/src/main/scala/com/evolutiongaming/kafka/journal/JsValueCodec.scala [hunk 1]
```diff
@@ -0,0 +1,49 @@
+package com.evolutiongaming.kafka.journal
+
+import com.evolutiongaming.jsonitertool.PlayJsonJsoniter
+import play.api.libs.json.{JsError, JsResult, JsSuccess, JsValue, Json}
+import scodec.bits.ByteVector
+
+import scala.util.{Failure, Success, Try}
+
+trait JsValueCodec {
```
Comment (t3hnar) [inline journal/src/main/scala/com/evolutiongaming/kafka/journal/JsValueCodec.scala]: Could you please divide them? Encode and Decode operations are in separate interfaces. 
Could you also go through all usages and use one or another

File: journal/src/main/scala/com/evolutiongaming/kafka/journal/Journals.scala [hunk 2]
```diff
@@ -78,7 +78,7 @@ object Journals {
     eventualJournal: EventualJournal[F],
     headCache: HeadCache[F],
     log: Log[F]
-  ): Journals[F] = {
+  )(implicit jsValueCodec: JsValueCodec): Journals[F] = {
```
Comment (t3hnar) [inline journal/src/main/scala/com/evolutiongaming/kafka/journal/Journals.scala]: could you please be consistent on naming things?

decide upon one name and not use interchangeably `jsCodec` and `jsValueCodec`

File: persistence/src/main/scala/akka/persistence/kafka/journal/KafkaJournal.scala [hunk 3]
```diff
@@ -102,6 +102,13 @@ class KafkaJournal(config: Config) extends AsyncWriteJournal { actor =>
     Resource.liftF(cassandraClusterOf)
   }
 
+  def jsValueCodec(config: KafkaJournalConfig): JsValueCodec = {
```
Comment (t3hnar) [inline persistence/src/main/scala/akka/persistence/kafka/journal/KafkaJournal.scala]: please make it F[JsValueCodec]

File: persistence/src/main/scala/akka/persistence/kafka/journal/KafkaJournal.scala [hunk 4]
```diff
@@ -102,6 +102,13 @@ class KafkaJournal(config: Config) extends AsyncWriteJournal { actor =>
...
+    config.serialization.jsonCodec match {
```
Comment (t3hnar) [inline persistence/src/main/scala/akka/persistence/kafka/journal/KafkaJournal.scala]: I'd like to be able to toggle that at runtime :)

File: persistence/src/main/scala/akka/persistence/kafka/journal/KafkaJournal.scala (hunk 3 above)
Comment (lavrov) [reply persistence/src/main/scala/akka/persistence/kafka/journal/KafkaJournal.scala]: I don't get how that would improve anything

File: journal/src/main/scala/com/evolutiongaming/kafka/journal/FromBytes.scala [hunk 5]
```diff
@@ -52,19 +47,11 @@ object FromBytes {
...
-      val jsValue = Json.parse(a.toArray)
-      JsSuccess(jsValue)
-    } catch {
-      case NonFatal(a) => JsError(s"failed to parse json $a")
-    }
-
-    val result = for {
-      a <- jsValue
-      a <- reads.reads(a)
+  def fromReads[F[_] : Monad : FromJsResult, A](implicit reads: Reads[A], decode: JsValueCodec.Decode[F]): FromBytes[F, A] = (a: ByteVector) => {
```
Comment (t3hnar) [inline journal/src/main/scala/com/evolutiongaming/kafka/journal/FromBytes.scala]: looks like line is too long

File: journal/src/main/scala/com/evolutiongaming/kafka/journal/JsValueCodec.scala [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+    },
+
+    decode = Decode { bytes =>
+      lift {
+        Try(Json.parse(bytes.toArray))
+      }
+    }
+  )
+
+  def jsoniter[F[_] : Applicative : FromTry]: JsValueCodec[F] = JsValueCodec(
```
Comment (t3hnar) [inline journal/src/main/scala/com/evolutiongaming/kafka/journal/JsValueCodec.scala]: @plokhotnyuk thanks for your work on jsoniter, just to let you know: we are going to use jsoniter in one of the most loaded part of our applications :)

File: persistence/src/main/scala/akka/persistence/kafka/journal/JournalAdapter.scala [hunk 7]
```diff
@@ -33,7 +33,11 @@ trait JournalAdapter[F[_]] {
 
 object JournalAdapter {
 
-  def of[F[_] : ConcurrentEffect : ContextShift : FromFuture : ToFuture : Parallel : Timer : LogOf : Runtime : RandomIdOf : FromGFuture : MeasureDuration : ToTry : FromTry : FromAttempt : FromJsResult : Fail](
+  def of[
+    F[_] : ConcurrentEffect : ContextShift : FromFuture : ToFuture : Parallel : Timer : LogOf : Runtime :
+    RandomIdOf : FromGFuture : MeasureDuration : ToTry : FromTry : FromAttempt : FromJsResult : Fail :
+    JsValueCodec.Encode : JsValueCodec.Decode
```
Comment (t3hnar) [inline persistence/src/main/scala/akka/persistence/kafka/journal/JournalAdapter.scala]: why not just : `JsValueCodec` ?

File: tests/src/test/scala/com/evolutiongaming/kafka/journal/IntegrationSuite.scala [hunk 8]
```diff
@@ -13,6 +13,7 @@ import com.evolutiongaming.kafka.journal.IOSuite._
 import com.evolutiongaming.scassandra.CassandraClusterOf
 import com.evolutiongaming.smetrics.{CollectorRegistry, MeasureDuration}
 import com.typesafe.config.ConfigFactory
+import JsValueCodec.Implicits.default
```
Comment (t3hnar) [inline tests/src/test/scala/com/evolutiongaming/kafka/journal/IntegrationSuite.scala]: this is a nice way to test whether this all works as expected, so please run it at least locally with jsoniter :)

File: journal/src/main/scala/com/evolutiongaming/kafka/journal/JsValueCodec.scala [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+    decode = Decode { bytes =>
+      lift {
+        PlayJsonJsoniter.deserialize(bytes.toArray)
+      }
+    }
+  )
+
+  object Implicits {
+
+    implicit def default[F[_] : Applicative : FromTry]: JsValueCodec[F] = JsValueCodec.playJson
```
[the remark this answers is not in the document]
Comment (lavrov) [reply journal/src/main/scala/com/evolutiongaming/kafka/journal/JsValueCodec.scala]: Default is only used  in test

[the remark this answers is not in the document]
Comment (t3hnar) [reply journal/src/main/scala/com/evolutiongaming/kafka/journal/JsValueCodec.scala]: so we can define it there in tests only ?

File: journal/src/main/scala/com/evolutiongaming/kafka/journal/JsValueCodec.scala (hunk 6 above)
Comment (lavrov) [reply journal/src/main/scala/com/evolutiongaming/kafka/journal/JsValueCodec.scala]: +1

File: persistence/src/main/scala/akka/persistence/kafka/journal/SerializationConfig.scala [hunk 10]
```diff
@@ -0,0 +1,24 @@
+package akka.persistence.kafka.journal
+
+import pureconfig.ConfigReader
+import pureconfig.generic.semiauto.{deriveEnumerationReader, deriveReader}
+
+final case class SerializationConfig(jsonCodec: SerializationConfig.JsonCoded)
```
[the remark this answers is not in the document]
Comment (t3hnar) [reply persistence/src/main/scala/akka/persistence/kafka/journal/SerializationConfig.scala]: could you please also add a test on this?

[the remark this answers is not in the document]
Comment (lavrov) [reply persistence/src/main/scala/akka/persistence/kafka/journal/SerializationConfig.scala]: done

**Pull request merged**